islandora.tokens.inc

```diff
@@ -95,8 +102,60 @@ function islandora_tokens($type, $tokens, array $data, array $options, Bubbleabl
           // alt is empty.
           $replacements[$original] = $alt;
           break;
+
+        case 'pdf_url':
+          $replacements[$original] = ' ' . islandora_url_to_service_file_media_by_mimetype($data['node'], 'application/pdf');
+          break;
       }
     }
   }
   return $replacements;
 }
+
+/**
+ * Gets Original File PDF file URL.
+ *
+ * @param object $node
+ *   A core drupal node object.
+ * @param string $mime_type
+ *   The name of the node's field to check for the specific relationship.
+ *
+ * @return string
+ *   The tokenized value for the given data.
+ */
+function islandora_url_to_service_file_media_by_mimetype($node, $mime_type) {
+  $islandora_utils = \Drupal::service('islandora.utils');
+  $origfile_term = $islandora_utils->getTermForUri('http://pcdm.org/use#OriginalFile');
+  $origfile_media = $islandora_utils->getMediaWithTerm($node, $origfile_term);
+  // Get the media file's mime_type value.
+  if (is_object($origfile_media)) {
+    $origfile_mime_type = ($origfile_media->hasField('field_mime_type')) ?
+      $origfile_media->get('field_mime_type')->getValue() : NULL;
+    $origfile_mime_type = (is_array($origfile_mime_type) &&
+      array_key_exists(0, $origfile_mime_type) &&
+      is_array($origfile_mime_type[0]) &&
+      array_key_exists('value', $origfile_mime_type[0])) ?
+      $origfile_mime_type[0]['value'] : '';
+    // Compare the media file's mime_type to the given value.
+    if ($origfile_mime_type == $mime_type) {
+      $vid = $origfile_media->id();
+      if (!is_null($vid)) {
+        $media = Media::load($vid);
+        $bundle = $media->bundle();
+        // Since this is Islandora and we assume the Original File is a
+        // Document type... but doing it dynamically.
+        $fid = $media->get('field_media_' . $bundle)->getValue();
+        $fid_value = (is_array($fid) && array_key_exists(0, $fid) &&
+          array_key_exists('target_id', $fid[0])) ?
+          $fid[0]['target_id'] : NULL;
+        if (!is_null($fid_value)) {
+          $file = File::load($fid_value);
+          if ($file) {
+            $url = $file->getFileUri();
```

Comment:
This returns files with a protocol of `public://` or `fedora://`, which can't be de-referenced from the outside. There's a function is IslanodraUtils you can use to get the download url of the file.

`$url = $islandora_utils->getDownloadUrl($file);` should do the trick.

Reply:
Confirmed, it definitely does the trick.  Make that change and this goes in :rocket: :bowing_man:

Reply:
getDownloadUrl replaced the previous code